### PR TITLE
Fix flaky CI tests (again 😓 )

### LINKED ISSUE #6659: Fix `test/test_gradio_component_cli.py:141 ` that breaks only on release PR
### Describe the bug

On every release PR, there's a backend test that fails because the JS packages have not been published yet. 

E.g. https://github.com/gradio-app/gradio/actions/runs/7093296419/job/19306412188?pr=6575

image

It'd be great if we can fix this so that we can rely on this test when releasing

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

https://github.com/gradio-app/gradio/actions/runs/7093296419/job/19306412188?pr=6575

### Screenshot

_No response_

### Logs

_No response_

### System Info

```shell
gradio==4.7.1
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description
~Looking at the audio and video functional tests since these are the only ones that flake for me.~

Things I've tried:
- Set n_workers to 1 in CI
- Rather than wait for upload to finish, wait for all network requests to finish. This is discouraged in the playwright docs for testing live production sites but I think in our case it's ok?
- Split up how we wait for play/pause events

Closes #6659

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: js/app/test/audio_component_events.spec.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ test("Audio click-to-upload uploads audio successfuly. File downloading works an
 	const uploader = await page.locator("input[type=file]");
 	await Promise.all([
 		uploader.setInputFiles(["../../test/test_files/audio_sample.wav"]),
-		page.waitForResponse("**/upload?*")
+		page.waitForResponse("**/upload*?*")
```
Comment (freddyaboulton) [inline js/app/test/audio_component_events.spec.ts:12]: Some had "**/upload?*" and some had "**/upload*?*". Using the latter just for consistency.

File: js/app/test/audio_component_events.spec.ts [hunk 2]
```diff
@@ -69,28 +69,26 @@ test("Play, Pause, and stop events work correctly.", async ({ page }) => {
...
+	const event_triggered = async (label: string) => {
+		const value = await page.getByLabel(label).inputValue();
+		expect(Number(value)).toBeGreaterThan(0);
+	};
+
 	await page
 		.getByTestId("waveform-Output Audio")
 		.getByLabel("Play", { exact: true })
 		.click();
+	await expect(async () => event_triggered("# Output Play Events")).toPass();
```
Comment (freddyaboulton) [inline js/app/test/audio_component_events.spec.ts:84]: I think what's happening is that we're clicking the pause button immediately after the play button and I think the server is not reacting as fast as the playwright client.

So what I did is wait for the play event to finish triggering. Then trigger the pause event. Same for video.

File: test/test_gradio_component_cli.py [hunk 3]
```diff
@@ -106,39 +106,35 @@ def test_show(capsys):
...
-
-    assert not (tmp_path / "frontend" / "node_modules").exists()
-    _install(tmp_path)
-    assert (tmp_path / "frontend" / "node_modules").exists()
+@pytest.mark.parametrize("template", ["Audio", "Image"])
+def test_build(template):
+    pnpm_lock = Path(__file__).parent / ".." / "pnpm-lock.yaml"
+    pnpm_copy = Path(__file__).parent / ".." / "pnpm-lock-copy.yaml"
+    shutil.copy(str(pnpm_lock), str(pnpm_copy))
+    # Using the js/preview/test directory will use the workspace code
```
Comment (freddyaboulton) [inline test/test_gradio_component_cli.py:114]: We'll always install the local `@gradio/` js packages so this should pass on release PRs

**Comment by abidlabs** [comment]
Thank you @freddyaboulton! In parallel I'll do a quick fix for the windows tests

File: .config/playwright.config.js [hunk 4]
```diff
@@ -10,5 +12,6 @@ export default {
 	},
 	testMatch: /.*.spec.ts/,
 	testDir: "..",
-	globalSetup: "./playwright-setup.js"
-};
+	globalSetup: "./playwright-setup.js",
+	workers: process.env.CI ? 1 : undefined
```
[the remark this answers is not in the document]
Comment (pngwn) [reply .config/playwright.config.js]: Isn't this always 1 in CI anyway?

File: js/app/test/audio_component_events.spec.ts (hunk 2 above)
Comment (pngwn) [reply js/app/test/audio_component_events.spec.ts:84]: Nice.

File: .config/playwright.config.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply .config/playwright.config.js]: It's 2 right now. I think we can actually keep.

File: js/app/test/image_component_events.spec.ts [hunk 5]
```diff
@@ -71,9 +64,7 @@ test("Image copy from clipboard dispatches upload event.", async ({ page }) => {
 		navigator.clipboard.write([new ClipboardItem({ [blob.type]: blob })]);
 	});
 
-	await page.pause();
```
Comment (freddyaboulton) [inline js/app/test/image_component_events.spec.ts:74]: How was this passing? It's supposed to block the entire script 🤔

File: js/app/test/chatinterface_streaming_echo.spec.ts [hunk 6]
```diff
@@ -9,51 +9,30 @@ test("chatinterface works with streaming functions and all buttons behave as exp
 	const clear_button = await page.getByRole("button", { name: "🗑️ Clear" });
 	const textbox = await page.getByPlaceholder("Type a message...");
 
-	let iterations: Promise<any>[] = [];
```
Comment (freddyaboulton) [inline js/app/test/chatinterface_streaming_echo.spec.ts:12]: How was this working before 🤔

File: gradio/cli/commands/components/build.py [hunk 7]
```diff
@@ -83,7 +83,7 @@ def _build(
             else:
                 live.update(":white_check_mark: Build succeeded!")
 
-        cmds = [sys.executable, "-m", "build", str(name)]
+        cmds = [shutil.which("python"), "-m", "build", str(name)]
```
Comment (freddyaboulton) [inline gradio/cli/commands/components/build.py:86]: Use the python in the current venv

File: gradio/cli/commands/components/publish.py [hunk 8]
```diff
@@ -1,11 +1,9 @@
-from __future__ import annotations
```
Comment (freddyaboulton) [inline gradio/cli/commands/components/publish.py:1]: This bug: https://github.com/tiangolo/typer/discussions/708

**Pull request merged**